File: 프로그래머스/lv3/43238. 입국심사/입국심사.py
def solution(n, times):
    # times.sort()
    start = 1
    end = times[-1]*n
    result = end

    # 여기서 mid가 심사 받는 데 걸리는 시간이 됨
    while start <= end:
        mid = (start+end) // 2
        # 30 / 7 = 4, 30 / 10 = 3 , 30분에 7명
        people = 0
        for time in times:
            people += (mid // time)

        if people < n:
            start = mid+1
        else:
        # 이 경우가 심사할 수 있는 사람 수가 n과 같거나 n보다 많기에
        # 결과에 반영해줘야 함
            end = mid-1
            result = min(result, mid)

    return result

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.Arrays;

class Solution {
    public long solution(int n, int[] times) {

        long left = 1; // 최소 시간
        long right = (long) n * getMax(times); // 최대 시간
        long answer = right;

        while (left <= right) {
            long mid = (left + right) / 2;
            if (canFinish(n, times, mid)) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return answer;
    }

    // 주어진 시간 내에 모든 사람을 심사할 수 있는지 판단하는 함수
    private boolean canFinish(int n, int[] times, long mid) {
        long total = 0;
        for (int time : times) {
            total += mid / time;
        }
        return total >= n;
    }

    // 심사 시간 배열에서 가장 큰 값을 반환하는 함수
    private long getMax(int[] times) {
        long max = times[0];
        for (int time : times) {
            if (time > max) {
                max = time;
            }
        }
        return max;
    }
}